src/Reserve.sol

```diff
@@ -4,30 +4,115 @@ pragma solidity ^0.8.7;
 import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
 import {IERC20} from "openzeppelin-contracts/token/ERC20/IERC20.sol";
 
+/// @notice The reserve interface as seen by the users.
 interface IReserve {
-    function withdraw(
+    /// @notice Deposits funds into the reserve.
+    /// The reserve will `transferFrom` `amt` tokens from the `from` address.
+    /// @param token The used token.
+    /// @param from The address from which funds are deposited.
+    /// @param amt The deposited amount.
+    function deposit(
         IERC20 token,
-        address to,
+        address from,
         uint256 amt
     ) external;
 
-    function deposit(
+    /// @notice Withdraws funds from the reserve.
+    /// The reserve will transfer `amt` tokens to the `to` address.
+    /// Only funds previously deposited can be withdrawn.
+    /// @param token The used token.
+    /// @param to The address to which funds are withdrawn.
+    /// @param amt The withdrawn amount.
+    function withdraw(
         IERC20 token,
-        address from,
+        address to,
         uint256 amt
     ) external;
 }
 
+/// @notice The reserve plugin interface required by the reserve.
+interface IReservePlugin {
+    /// @notice Called by the reserve immediately after transferring funds for deposition.
+    /// @param token The used token.
+    /// @param amt The amount which has been transferred for deposition.
+    function afterDeposition(IERC20 token, uint256 amt) external;
+
+    /// @notice Called by the reserve right before transferring funds for withdrawal.
+    /// The reserve will `transferFrom` the tokens from the plugin address.
+    /// The reserve can always withdraw everything that has been ever deposited, but never more.
+    /// @param token The used token.
+    /// @param amt The amount which will be transferred.
+    function beforeWithdrawal(IERC20 token, uint256 amt) external;
+}
+
+/// @notice The ERC-20 tokens reserve contract.
+/// The registered users can deposit and withdraw funds.
+/// The reserve by default doesn't do anything with the tokens,
+/// but for each ERC-20 address a plugin can be registered for tokens storage.
 contract Reserve is IReserve, Ownable {
+    /// @notice The dummy plugin address meaning that no plugin is being used.
```

Comment:
I am wondering if we are completely open to support any ERC20 or if governance should have a potential `blackList` mapping.

If we want such a blacklist we potentially could manage it here directly. 

`deposit` and `withdraw` would just revert, if a token is on a blacklist.

Reply:
I think that we should just have a `BlacklistPlugin`, which implements the plugin API, but reverts for whichever token it's assigned. It's much cleaner and more powerful, because it makes it possible to e.g. allow refunding of tokens already in the system.

Reply:
Thats, a good idea. :+1:

Reply:
That made me thinking that maybe the plugin API should implement a lifecycle for plugins making it easy to gracefully set them up and then shut down :thinking: E.g. implementation of the `BlacklistPlugin` would become trivial with such interface. 

```solidity
interface IReservePlugin {
    // Unchanged
    function afterDeposition(IERC20 token, uint256 amt) external;
    function beforeWithdrawal(IERC20 token, uint256 amt) external;

    // Acts like `afterDeposition`, but called when setting a plugin after the initial deposition
    function afterSetting(IERC20 token, uint256 amt) external;

    // Acts like `beforeWithdrawal`, but called before unsetting and draining the plugin
    function beforeUnsetting(IERC20 token, uint256 amt) external;
}
```

Reply:
Yes, that's nicer. Otherwise, `afterDeposition` would have to handle the first deposit as an edge case.

Normally, the first step might require some initial steps.

```solidity
    function start(IERC20 token, uint256 amt) external;
    function shutdown(IERC20 token, uint256 amt) external;
}
```

I think from the naming the plugin is not aware that it is `set`. It can `start` and `shutdown` with a specific balance.

You get my naming philosophy. :)

Reply:
The lower layer naming should not reflect the higher layer specific purpose.

Reply:
Added